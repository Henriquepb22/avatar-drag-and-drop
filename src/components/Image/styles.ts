import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
    ${({ theme }) => css`
        display: flex;
        align-items: center;
        justify-content: center;
        background: ${theme.colors.grayMedium};
        border-radius: 50%;
        width: 11.4rem;
        height: 11.4rem;
        overflow: hidden;
    `}
`

type ImageProps = {
    zoomLevel: number
}

export const Image = styled.img<ImageProps>`
    ${({ zoomLevel }) => css`
        max-width: 24rem;
        max-height: 24rem;
        transform: scale(${zoomLevel});
    `}
`
