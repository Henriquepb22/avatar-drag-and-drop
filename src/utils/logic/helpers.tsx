const isImageFile = (file: File) => {
    const fileType = file.type
    const baseMimeType = fileType.split('/')[0]

    if (baseMimeType !== 'image') return false
    return true
}

export const getFileDataUrl = async (file: File): Promise<string> => {
    const reader = new FileReader()

    return new Promise((resolve, reject) => {
        const isImageType = isImageFile(file)

        if (isImageType) {
            reader.readAsDataURL(file)
            reader.onload = () => {
                if (reader.result) {
                    resolve(reader.result.toString())
                }
            }
        } else {
            reader.abort()
            reject(new Error('File is not an recognized image type'))
        }
    })
}

export const cropImage = async (
    imageSrc: string,
    zoomLevel: number
): Promise<string> => {
    const canvas = document.createElement('canvas')
    const context = canvas.getContext('2d')
    const image = new Image()

    return new Promise((resolve, reject) => {
        image.src = imageSrc

        image.onload = () => {
            if (context) {
                const needCompression =
                    Math.max(canvas.width, canvas.height) > 1000

                canvas.width = needCompression ? image.width / 2 : image.width
                canvas.height = needCompression
                    ? image.height / 2
                    : image.height

                context.drawImage(image, 0, 0, canvas.width, canvas.height)
                context.globalCompositeOperation = 'destination-in'
                context.beginPath()
                context.arc(
                    canvas.width * 0.5,
                    canvas.height * 0.5,
                    (Math.max(canvas.width, canvas.height) * 0.5) /
                        (zoomLevel * 2),
                    0,
                    2 * Math.PI
                )
                context.fill()
                context.globalCompositeOperation = 'source-over'

                resolve(canvas.toDataURL('image/png', 1))
            }
        }

        image.onerror = () => {
            reject(new Error('Something bad happened...'))
        }
    })
}
