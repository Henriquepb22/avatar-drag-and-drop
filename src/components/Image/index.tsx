import * as S from './styles'

type ImageProps = {
    src: string
    altText: string
    zoomLevel?: number
}

const Image = ({ src, altText, zoomLevel = 1 }: ImageProps) => {
    return (
        <S.Wrapper>
            <S.Image src={src} alt={altText} zoomLevel={zoomLevel} />
        </S.Wrapper>
    )
}

export default Image
