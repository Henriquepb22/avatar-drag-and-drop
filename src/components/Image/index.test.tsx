import { renderWithTheme } from 'utils/tests/helpers'
import { screen } from '@testing-library/dom'

import Image from '.'

describe('<Image />', () => {
    it('should render image correctly', () => {
        const { container } = renderWithTheme(
            <Image src="image.png" altText="some image" />
        )

        expect(
            screen.getByRole('img', { name: /some image/i })
        ).toBeInTheDocument()
        expect(container.firstChild).toMatchSnapshot()
    })

    it('should change image scale by zoom level', () => {
        renderWithTheme(
            <Image src="image.png" altText="some image" zoomLevel={1.5} />
        )

        const image = screen.getByRole('img', { name: /some image/i })
        expect(image).toHaveStyle({
            transform: 'scale(1.5)',
        })
    })
})
