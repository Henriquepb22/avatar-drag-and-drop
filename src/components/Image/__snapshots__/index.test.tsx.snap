// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> should render image correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #C3CBD5;
  border-radius: 50%;
  width: 11.4rem;
  height: 11.4rem;
  overflow: hidden;
}

.c1 {
  width: 22.8rem;
  height: 22.8rem;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  object-fit: cover;
}

<div
  class="c0"
>
  <img
    alt="some image"
    class="c1"
    src="image.png"
  />
</div>
`;
